File: 0211-design-add-and-search-words-data-structure/0211-design-add-and-search-words-data-structure.java
class TrieNode {
    HashMap<Character ,TrieNode> children = new HashMap<>();
    boolean isWord = false;
}

class WordDictionary {

    TrieNode root;

    public WordDictionary() {
        root = new TrieNode();
    }
    
    public void addWord(String word) {
        TrieNode current = root;
        for(char ch: word.toCharArray()){
            current.children.putIfAbsent(ch, new TrieNode());
            current = current.children.get(ch);
        }
        current.isWord = true;
    }

    public boolean dfs(String word, int j, TrieNode root){
        TrieNode current = root;
        for(int i=j; i<word.length();i++){
            char ch = word.charAt(i);
            if(ch=='.'){
                for(TrieNode child: current.children.values()){
                    if (dfs(word, i+1, child)) return true;
                }
                return false;
            }
            else{
                if(!current.children.containsKey(ch)){
                    return false;
                }
                current = current.children.get(ch);
            }
        }
        return current.isWord;
    }
    
    public boolean search(String word) {
        return dfs(word, 0, root);
    }
}

/**
 * Your WordDictionary object will be instantiated and called as such:
 * WordDictionary obj = new WordDictionary();
 * obj.addWord(word);
 * boolean param_2 = obj.search(word);
 */

File: 0211-design-add-and-search-words-data-structure/0211-design-add-and-search-words-data-structure.py
class TrieNode:
    def __init__(self):
        self.children = {}
        self.isWord = False

class WordDictionary:

    def __init__(self):
        self.root = TrieNode() 

    def addWord(self, word: str) -> None:
        cur = self.root
        for c in word:
            if c not in cur.children:
                cur.children[c] = TrieNode()
            cur = cur.children[c]
        cur.isWord = True

    def search(self, word: str) -> bool:
        def dfs(i, root):
            cur = root
            for j in range(i, len(word)):
                ch = word[j]
                if ch == '.':
                    for child in cur.children.values():
                        if dfs(j + 1, child):
                            return True
                    return False
                else:
                    if ch not in cur.children:
                        return False
                    cur = cur.children[ch]
            return cur.isWord
        return dfs(0, self.root)

    

# Your WordDictionary object will be instantiated and called as such:
# obj = WordDictionary()
# obj.addWord(word)
# param_2 = obj.search(word)